### PR TITLE
accuracy: support ignore_label

### PR DESCRIPTION
**Pull request by henry0312**
Some users want to calculate accuracy except a certain label.
For example, we usually don't obtain accuracy for `<EOS>` label under a task of predicting next charcters. 

Thank you.

### CODE REVIEW

File: chainer/functions/evaluation/accuracy.py [hunk 1]
```diff
@@ -25,22 +28,35 @@ def forward(self, inputs):
...
         y = y.reshape(len(y), -1)  # flatten
-        pred = y.argmax(axis=1)
-        return xp.asarray((pred == t).mean(dtype='f')),
+
+        if self.ignore_label is not None:
+            mask = (t == self.ignore_label)
+            ignore_cnt = mask.sum()
+
+            # will always be true when the true label is ignore_label
+            pred = xp.where(mask, self.ignore_label, y.argmax(axis=1))
```
Comment (henry0312) [inline chainer/functions/evaluation/accuracy.py]: If `cupy.where` returns indexes, we could make the code better.

Comment (henry0312) [inline chainer/functions/evaluation/accuracy.py]: Also, we would need [Advanced Indexing](http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#advanced-indexing "Advanced Indexing")

Comment (unnonouno) [inline chainer/functions/evaluation/accuracy.py]: Please write a TODO comment about it.

File: chainer/functions/evaluation/accuracy.py [hunk 2]
```diff
@@ -25,22 +28,35 @@ def forward(self, inputs):
...
+            count = (pred == t).sum() - ignore_cnt
+            total = len(t) - ignore_cnt
+
+            return xp.asarray(float(count)/total, dtype='f'),
```
Comment (unnonouno) [inline chainer/functions/evaluation/accuracy.py]: Please insert spaces beside `/`.

Comment (unnonouno) [inline chainer/functions/evaluation/accuracy.py]: Check if `total > 0`. When `total == 0`, please return `0`.

File: tests/chainer_tests/functions_tests/evaluation_tests/test_accuracy.py [hunk 3]
```diff
@@ -11,26 +11,43 @@
...
 
+@testing.parameterize(
+    {'x_data': numpy.random.uniform(-1, 1, (10, 3)).astype(numpy.float32),
+     't_data': numpy.random.randint(3, size=(10,)).astype(numpy.int32),
+     'ignore_label': None},
+    {'x_data': numpy.random.uniform(-1, 1, (20, 3)).astype(numpy.float32),
+     't_data': numpy.random.randint(3, size=(20,)).astype(numpy.int32),
+     'ignore_label': 0},
+)
 class TestAccuracy(unittest.TestCase):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_accuracy.py]: Please add a test case that `t_data` only contains `ignore_label`.

**Comment by unnonouno** [comment]
Thank you for a PR. I wrote some comments.

**Comment by henry0312** [comment]
updated

**Comment by unnonouno** [comment]
Great work!! The code is OK. Please check the result on Travis CI. Fix code style. https://travis-ci.org/pfnet/chainer/jobs/130480117

You can check it with hacking package in your local environment. https://pypi.python.org/pypi/hacking

**Comment by henry0312** [comment]
updated again

File: chainer/functions/evaluation/accuracy.py [hunk 4]
```diff
@@ -25,22 +28,42 @@ def forward(self, inputs):
...
         y = y.reshape(len(y), -1)  # flatten
-        pred = y.argmax(axis=1)
-        return xp.asarray((pred == t).mean(dtype='f')),
+
+        if self.ignore_label is not None:
+            mask = (t == self.ignore_label)
+            ignore_cnt = mask.sum()
+
+            # will always be true when the true label is ignore_label
+            # TODO(someone)
```
Comment (unnonouno) [inline chainer/functions/evaluation/accuracy.py]: Please write your name or your account name. This name is not who need todo, but who wrote the TODO. *We* will fix it.

**Comment by unnonouno** [comment]
This is the last comment!

**Comment by henry0312** [comment]
updated!

**Comment by unnonouno** [comment]
Looks great! Thank you very much.

**Comment by henry0312** [comment]
thank you for your review.

**Pull request merged**